File: src/test/scala/com/matchilling/playground/scala/RecursionTest.scala
package com.matchilling.playground.scala

import java.io.{ByteArrayOutputStream, PrintStream}

import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.{AfterEach, BeforeEach, Test}

class RecursionTest {
  final private val outContent = new ByteArrayOutputStream
  final private val originalOut = System.out

  @BeforeEach
  def setUpStreams(): Unit = {
    System.setOut(new PrintStream(outContent))
  }

  @AfterEach
  def restoreStreams(): Unit = {
    System.setOut(originalOut)
  }

  @Test
  def countDown(): Unit = {
    Recursion.countDown(5)
    val expectedOutput = "5\n" + "4\n" + "3\n" + "2\n" + "1\n"
    assertEquals(expectedOutput, outContent.toString)
  }

  @Test
  def countUp(): Unit = {
    Recursion.countUp(5)
    val expectedOutput = "1\n" + "2\n" + "3\n" + "4\n" + "5\n"
    assertEquals(expectedOutput, outContent.toString)
  }
}

File: src/test/java/com/matchilling/playground/java/RecursionTest.java
package com.matchilling.playground.java;

import org.junit.jupiter.api.AfterEach;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

import java.io.ByteArrayOutputStream;
import java.io.PrintStream;

import static org.junit.jupiter.api.Assertions.assertEquals;

class RecursionTest {

    private final ByteArrayOutputStream outContent = new ByteArrayOutputStream();
    private final PrintStream originalOut = System.out;

    @BeforeEach
    public void setUpStreams() {
        System.setOut(new PrintStream(outContent));
    }

    @AfterEach
    public void restoreStreams() {
        System.setOut(originalOut);
    }

    @Test
    void countDown() {
        Recursion.countDown(5);

        String expectedOutput = "5\n4\n3\n2\n1\n";

        assertEquals(expectedOutput, outContent.toString());
    }

    @Test
    void countUp() {
        Recursion.countUp(5);

        String expectedOutput = "1\n2\n3\n4\n5\n";

        assertEquals(expectedOutput, outContent.toString());
    }
}